File: src/main/java/graph/algorithm/GraphPath.java
package graph.algorithm;

import graph.edge.Edge;
import graph.vertex.Vertex;
import org.jetbrains.annotations.NotNull;

import java.util.ArrayList;
import java.util.Iterator;
import java.util.Spliterator;
import java.util.function.Consumer;
import java.util.stream.Collectors;

public final class GraphPath implements Iterable<Edge> {
    private final ArrayList<Edge> links;
    private final boolean cycle;

    public GraphPath() {
        this(false);
    }

    public GraphPath(boolean cycle) {
        this.cycle = cycle;
        links = new ArrayList<>();
    }

    public GraphPath add(Edge edge) {
        links.add(edge);
        return this;
    }

    public GraphPath clear() {
        links.clear();
        return this;
    }

    public GraphPath remove() {
        if (!links.isEmpty())
            links.remove(links.size() - 1);
        return this;
    }

    public int getLength() {
        return links.size();
    }

    public int getWeight() {
        return links.stream().mapToInt(Edge::getWeight).sum();
    }

    public boolean exists() {
        return cycle || !links.isEmpty();
    }

    @NotNull
    @Override
    public Iterator<Edge> iterator() {
        return links.iterator();
    }

    @Override
    public void forEach(Consumer<? super Edge> action) {
        links.forEach(action);
    }

    @Override
    public Spliterator<Edge> spliterator() {
        return links.spliterator();
    }

    @Override
    public String toString() {
        if (!exists())
            return "No path";
        if (links.isEmpty())
            return "Cycle";
        return "[" + getWeight() + "] " + getFrom() + ">" +
                links.stream()
                .map(Edge::getToNode)
                .map(Object::toString)
                .collect(Collectors.joining(">"));
    }

    public Vertex getFrom() {
        return links.get(0).getFromNode();
    }

    public Vertex getTo() {
        return links.get(links.size() - 1).getToNode();
    }

    public boolean ofNodes(Vertex from, Vertex to) {
        if (from == to && cycle)
            return true;
        if (links.isEmpty())
            return false;
        return getFrom().equals(from) &&
                getTo().equals(to);
    }

    // sending package in datagram mode
    public boolean close(int weight) {
        int sum = 0;
        boolean result = false;
        for (int i = links.size() - 1; i >= 0; --i) {
            Edge edge = links.get(i);
            if (weight > sum) {
                sum += edge.getWeight();
            } else {
                edge.setUsed(false);
                result = true;
            }
        }
        return result;
    }
}

File: src/main/java/mvc/model/algorithms/GraphPath.kt
package mvc.model.algorithms

import mvc.model.channel.Channel
import mvc.model.node.Node

import java.util.ArrayList
import java.util.Spliterator
import java.util.function.Consumer

class GraphPath @JvmOverloads constructor(private val cycle: Boolean = false, private val links: ArrayList<Channel> = ArrayList()) : Iterable<Channel> {

    val length: Int
        get() = links.size

    val weight: Int
        get() = links.sumBy(Channel::weight)

    val from: Node
        get() = links[0].fromNode

    val to: Node
        get() = links[links.size - 1].toNode

    fun add(channel: Channel): GraphPath {
        links.add(channel)
        return this
    }

    fun clear(): GraphPath {
        links.clear()
        return this
    }

    fun remove(): GraphPath {
        if (!links.isEmpty())
            links.removeAt(links.size - 1)
        return this
    }

    fun exists(): Boolean {
        return cycle || !links.isEmpty()
    }

    override fun iterator(): Iterator<Channel> {
        return links.iterator()
    }

    override fun forEach(action: Consumer<in Channel>) {
        links.forEach(action)
    }

    override fun spliterator(): Spliterator<Channel> {
        return links.spliterator()
    }

    override fun toString(): String {
        if (!exists())
            return ""
        if (links.isEmpty())
            return "Цикл"
        val base = StringBuilder()
        base.append("Шлях (")
                .append(to)
                .append("): транзитних ділянок ")
                .append(length)
                .append(", загальна вага ")
                .append(weight)
                .append(" [")
                .append(links[0].fromNode)
                .append(" -> ")
        return base.toString() + links
                .map(Channel::toNode)
                .joinToString(" -> ") + "]"
    }

    fun ofNodes(from: Node, to: Node): Boolean {
        if (from === to && cycle)
            return true
        return if (links.isEmpty()) false else from == from && to == to
    }

    // sending package in datagram mode
    fun close(weight: Int): Boolean {
        var sum = 0
        for (i in links.indices.reversed()) {
            val channel = links[i]
            if (weight > sum) {
                sum += channel.weight
            } else {
                channel.isUsed = false
                return true
            }
        }
        return false
    }
}
